Flaskapi merge

SavSanta opened the pull request:
# Description

This pull requests implements a functional API backend using the Flask's builtin test webserver, a test SQLite3 database†, and a DOCKERFILE that can be used to build and run an image using alpine linux.

There are few routes for calls established and can be interfaced with  which is briefly documented at the main index page once the server is stood up as well as on this github branch documentation using the `curl` program (available typically on Linux/Unix/Mac. Windows users should be able to similarly utilize something like Invoke-WebRequest).


† Note that the SQLite3 database backend files are not attached as it is meant for usage as testing and we're still missing one contributor's code for NTEE normalized names. It is attached for usage here if you wish to download and add it to the project directory.

Commit:
Delete server.py

Commit:
Delete requirements.txt

Commit:
app initial

Commit:
add api folder

Commit:
mod Models field types and api route for all orgs

Commit:
add marshmallow for JSON serialization to requirements

Commit:
add import of model to api/views.py for debug test

Commit:
Change Models.py to Marshmallow types. Adjust init.py to instantiate Marshmallow

Commit:
Change Models.py amount fields for Asset, Income, and Revenue to Integer types

Commit:
adjust spacing

Commit:
rename get_activities to get_activity

Commit:
add a field to models.py for RAW STREET sql field

Commit:
add a fields to models.py for geocodes and latitude/longitude fields

Commit:
Change geographics coordinates to FLOAT types in models.py

Commit:
prevent sorting of JSON response object

Commit:
Models.py updated with REPR dunder

Commit:
Change models default amount andd nullables

Commit:
Set JSON_SORT_KEYS to false and remove unused forms

JSON sorting should possibly be set to false in order to allow a custom returned JSON object that makes sense from the API. \n\n. Removed currently unused forms keyword paramater so for live debug testing via flask shell

Commit:
Signature changed to remove forms=forms

Actually removed currently unused forms keyword paramater so for live debug testing via flask shell

Commit:
API route changed to /all

Changed API routes to /all and modified code to currently return a mock simulation of all Nonprofit objects. There now exists a NonprofitSchema(many=True) in order to retrieve multiple serialized Nonprofit objects and serialize them with JSONify

Commit:
Merge web branch 'flaskapi' that includes config paramaters

Commit:
Models.py updates to order and existing database

Added "__table_args__ = {'extend_existing': True}" to work with existing test sqlite3 database. Can probably be removed when starting fresh
Added "Ordered = True" option to return a consistent JSON format in the order it was serialized which should match same as CSV source header order. This is mainly for human convenience of course.

Commit:
Add API routes for Address returns

Added to API routes that use getattr to return data from the model for Nonprofit addresses

Commit:
Add API routes for Geocode (Lat/Long coord) returns

Added API routes to view using getattr to return data from the model for Nonprofit Longitude and Latitude coordinates.

Commit:
Dockerfile updates for running flask test server containerized

Commit:
Refactor dockerfile completely.

Refactor the dockerfile to use ENTRYPOINT and COPY

Commit:
Update README.md

Commit:
update requirements.txt with version numbers

Commit:
Update README.md

Commit:
add 0.0.0.0 to '__main__' runs 

Saves time (instead of running flask run --host format) but mainly lets the lets the dockerfile image and containers operate properly work .

Commit:
Update README.md

Commit:
Function-ify the JSON response sifting filter for Model

Create function that helps filter for the custom selected attributes from the database Model/Schema that we define.

Commit:
Update README.md

Commit:
Update README.md

Commit:
Remove trailing slashes for unique URLs

Change canonical URLs to omit the slash. This way the API must be hit strictly on the path rather than offer a redirect as it does with the trailing slash. (ie 404 is better than 302s)

Commit:
Update README.md

Commit:
Change template signature

Commit:
update home.html

Commit:
fillin example list items in html

Commit:
Update home.html

SavSanta commented:
SQLite3 TEST Database -- 
[database-devel.sqlite3.zip](https://github.com/CodeForBaltimore/CitySync/files/4295282/database-devel.sqlite3.zip)

stoopidJSON commented on README.md:
```diff
@@ -1,42 +1,35 @@
+*  [Flask](https://palletsprojects.com/p/flask/)
+*  [Flask-Marshmallow](https://flask-marshmallow.readthedocs.io)
+*  [Flask-SQLAlchemy](https://flask-sqlalchemy.palletsprojects.com)
+*  [Docker](https://docker.com) (*for optional container image usage*)
```
Maybe add a small bit about adding the SQL db

SavSanta replied:
The "test" database is not intended to be shipped. This is for dev testing only. Someone will need to figure out the proper seeding mechanism required once someone moves to an official database. Hopefully, no one is intending to run off a SQLITE3 on an actual production system due to concurrent writes causing DBLocks

stoopidJSON replied:
Fair enough, but it is a requirement to run the application at this time. I would just suggest adding something to explain that for now, but that's your call.

stoopidJSON commented on citysync/requirements.txt:
```diff
@@ -0,0 +1,25 @@
+MarkupSafe==1.1.1
+marshmallow==3.3.0
+marshmallow-sqlalchemy==0.21.0
+pkg-resources==0.0.0
```
This is causing an error during Docker build `ERROR: Could not find a version that satisfies the requirement pkg-resources==0.0.0`

SavSanta replied:
Thatcan be removed it was generated via `pip freeze` which I guess added all self-contained dependencies. I just ran it on an EC2 instance and it works with that line removed

stoopidJSON reviewed (changes requested):
I'm not able to run this yet due to the error noted in the comments, but this looks great so far!

Commit:
Remove pip freeze generated pkgsource

Virtualenvs on Ubuntu platform creates a `pkg-resources==0.0.0` in output of pip freeze that stops docker build command. Removing it fixes this.

stoopidJSON reviewed (approved):
A minor README addition could be added, but otherwise, it runs locally. Looks good to me!
